### PR TITLE
docs: fix return type description of Trainer.validate/test

### LINKED ISSUE #7486: [Bug] Trainer.validate() incorrect return-type (same for test())
## 🐛 Bug

- [ ] FIXME 1: with `model.validation_epoch_end = None` the `validate()` does not return `[validation_step()s]`
- [ ] FIXME 2: with `assert model.validation_epoch_end is not None` validate() does not return `{}`, but `[{}]`, where `len([]) == 1` (depends on num dataloaders??)
  - [ ] mention in doc that the len of returned `[{}]` depends on number of dataloaders?
- [x] TODO shouldn't doc on return-type for `test()` and `validate()` be similar/same? https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html#test 

## Please reproduce using the BoringModel

```python
import os
import torch
from torch.utils.data import DataLoader, Dataset
from pytorch_lightning import LightningModule, Trainer

class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)
    def __getitem__(self, index):
        return self.data[index]
    def __len__(self):
        return self.len

class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
    def forward(self, x):
        return self.layer(x)
    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}
    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)
        return {'valid_loss': loss, 'other': 42}
    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)
    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=5,
        limit_val_batches=3,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
    )
    # pl.Trainer.validate() -> return-type?
    # > Returns
    #     The dictionary with final validation results returned by validation_epoch_end. If validation_epoch_end is not defined, the output is a list of the dictionaries returned by validation_step.
    # from: https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html?highlight=validate#validate
    #       ... and https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html?highlight=validation_epoch_end#validation-epoch-end 
    
    #FIXME 1: with `model.validation_epoch_end = None` validate() does not return [validation_step()s]
    #FIXME 2: with `assert model.validation_epoch_end is not None` validate() does not return {}, but [{}], where len([]) == 1 (depends on num dataloaders??)
    #TODO shouldn't doc on return-type for test() and validate() be similar/same? https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html#test 
    
    
    # case 1
    # un/comment as needed - disable hooks mentioned in validate() & test() doc to change behavior of return-type
    # model.on_validation_epoch_end = None
    # model.validation_epoch_end = None
    # model.on_train_epoch_end = None
    # model.training_epoch_end = None
    # model.on_test_epoch_end = None
    # model.test_epoch_end = None
    
    # case 2
    trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)
    val = trainer.validate(model, val_dataloaders=val_data)
    tst = trainer.test(model, test_dataloaders=test_data)

if __name__ == '__main__':
    run()
```

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !
PL1.3 from PyPI

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### Additional context

Discussed on slack: https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1620748519201000

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #7486

Contains a minor update to typehints as well.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
